qcodes/utils/wrappers.py

```diff
@@ -336,70 +347,75 @@ def _do_measurement(loop: Loop, set_params: tuple, meas_params: tuple,
+            except (ClosedError, ConnectionError):
+                log.warning('Remote process crashed png will not be saved')
         else:
-            _ = loop.run(use_threads=use_threads)
-    except KeyboardInterrupt:
-        interrupted = True
-        print("Measurement Interrupted")
-    if do_plots:
-        # Ensure the correct scaling before saving
+            plot = None
         try:
-            plot.autorange()
-            plot.save()
-        except (ClosedError, ConnectionError):
-            log.warning('Remote process crashed png will not be saved')
-        plt.ioff()
-        pdfplot, num_subplots = _plot_setup(data, meas_params, useQT=False)
-        # pad a bit more to prevent overlap between
-        # suptitle and title
-        pdfplot.rescale_axis()
-        pdfplot.fig.tight_layout(pad=3)
-        title_list = plot.get_default_title().split(sep)
-        title_list.insert(-1, CURRENT_EXPERIMENT['pdf_subfolder'])
-        title = sep.join(title_list)
+            if do_plots:
+                _ = loop.with_bg_task(plot.update).run(use_threads=use_threads)
+            else:
+                _ = loop.run(use_threads=use_threads)
+        except KeyboardInterrupt:
+            interrupted = True
+            print("Measurement Interrupted")
+        if do_plots:
+            # Ensure the correct scaling before saving
+            try:
+                plot.autorange()
+                plot.save()
+            except (ClosedError, ConnectionError):
+                log.warning('Remote process crashed png will not be saved')
+            plt.ioff()
+            pdfplot, num_subplots = _plot_setup(data, meas_params, useQT=False)
+            # pad a bit more to prevent overlap between
+            # suptitle and title
+            pdfplot.rescale_axis()
+            pdfplot.fig.tight_layout(pad=3)
+            title_list = plot.get_default_title().split(sep)
+            title_list.insert(-1, CURRENT_EXPERIMENT['pdf_subfolder'])
+            title = sep.join(title_list)
+
+            pdfplot.save("{}.pdf".format(title))
+            if pdfdisplay['combined'] or (num_subplots == 1 and pdfdisplay['individual']):
+                pdfplot.fig.canvas.draw()
+                plt.show()
+            else:
+                plt.close(pdfplot.fig)
+            if num_subplots > 1:
+                _save_individual_plots(data, meas_params, pdfdisplay['individual'])
+            plt.ion()
+        if CURRENT_EXPERIMENT.get('device_image'):
+            log.debug('Saving device image')
+            save_device_image(tuple(sp[0] for sp in set_params))
+
+        # add the measurement ID to the logfile
+        with open(CURRENT_EXPERIMENT['logfile'], 'a') as fid:
+            print("#[QCoDeS]# Saved dataset to: {}".format(data.location),
+                  file=fid)
+        if interrupted:
+            raise KeyboardInterrupt
+    except:
+        log.exception("Exception in doND")
```

Comment:
I think the documentation is a bit vague, so I'll ask upfront: is it true that `log.exception` automatically logs information about the exception so that there is no need to do `except Exception as e` and then `log.exception('This happened: {}'.format(e))` ?

Reply:
Yes that is the point of using log.exception. It will automatically add the traceback unlike the default of the other logging functions. Otherwise it is at the same level as ERROR. 
For a keyboard interupt it would be 

```
2017-11-21 09:16:05,492 - qcodes.utils.wrappers - ERROR - Exception in doND
Traceback (most recent call last):
  File "/Users/jhn/src/Qcodes-dk/qcodes/utils/wrappers.py", line 414, in _do_measurement
    raise KeyboardInterrupt
KeyboardInterrupt
```
or if you try to divide by zero within the measurement function
```
2017-11-20 16:04:39,746 - qcodes.utils.wrappers - ERROR - Exception in doND
Traceback (most recent call last):
  File "/Users/jhn/src/Qcodes-dk/qcodes/utils/wrappers.py", line 369, in _do_measurement
    1/0
ZeroDivisionError: division by zero
```

Reply:
That makes sense, thanks. I think this looks good.